File: ingestion-service/src/main/scala/datainsider/ingestion/domain/ClickhouseSource.scala
package datainsider.ingestion.domain

import com.twitter.finatra.validation.constraints.NotEmpty

import java.net.URI

/**
  * created 2022-07-21 5:02 PM
  *
  * @author tvc12 - Thien Vi
  */
object ClickhouseSource {
  implicit class ClickhouseSourceImplicit(val value: ClickhouseSource) extends AnyVal {

    def getHost(): String = {
      getUri().getHost
    }

    private def getUri(): URI = {
      URI.create(value.jdbcUrl.replace("jdbc:", ""))
    }

    /**
      * Get the port of from jdbc url, default port is 8123
      */
    def getPort(): String = {
      val port: Int = getUri().getPort
      if (port == -1) {
        "8123"
      } else {
        port.toString
      }
    }
  }
}

@SerialVersionUID(20220720L)
case class ClickhouseSource(
    @NotEmpty jdbcUrl: String,
    @NotEmpty username: String,
    password: String,
    clusterName: String
)

File: client/di_core/ClickhouseConfig/Domain/ClickhouseSource/ClickhouseSource.ts
export class ClickhouseSource {
  jdbcUrl: string;
  username: string;
  password: string;
  clusterName: string;

  constructor(jdbcURL: string, username: string, password: string, clusterName: string) {
    this.jdbcUrl = jdbcURL;
    this.username = username;
    this.password = password;
    this.clusterName = clusterName;
  }

  static fromObject(obj: any): ClickhouseSource {
    return new ClickhouseSource(obj.jdbcUrl, obj.username, obj.password, obj.clusterName);
  }

  static default() {
    return new ClickhouseSource('', '', '', '');
  }
}
